corehq/apps/user_importer/helpers.py

```diff
@@ -9,3 +16,269 @@ def spec_value_to_boolean_or_none(user_spec_dict, key):
+        if self.is_new_user:
+            return
+        for name, new_value in changes.items():
+            if self.original_user_doc[name] != new_value:
+                self.fields_changed[name] = new_value
+                self._save = True
+
+    def add_change_message(self, message):
+        # ignore for new user since the whole user doc is logged for a new user
+        if self.is_new_user:
+            return
+        self.messages.append(message)
+        self._save = True
+
+    def add_info(self, info):
+        # useful info for display like names
+        self.messages.append(info)
+        self._save = True
+
+    def save(self):
+        if self.is_new_user or self._save:
+            action = UserModelAction.CREATE if self.is_new_user else UserModelAction.UPDATE
+            fields_changed = None if self.is_new_user else self.fields_changed
+            log_user_change(
+                self.domain,
+                self.user,
+                changed_by_user=self.changed_by_user,
+                changed_via=self.changed_via,
+                message=". ".join(self.messages),
+                action=action,
+                fields_changed=fields_changed
+            )
+
+
+class BaseUserImporter(object):
+    """
+    Imports a Web/CommCareUser via bulk importer and also handles the logging
+    save_log should be called explicitly to save logs, after user is saved
+    """
+    def __init__(self, upload_domain, user_domain, user, upload_user, is_new_user, via):
+        self.user_domain = user_domain
+        self.user = user
+        self.upload_user = upload_user
+        self.logger = UserChangeLogger(upload_domain, user=user, is_new_user=is_new_user,
+                                       changed_by_user=upload_user, changed_via=via)
+
+        self.role_updated = False
+
+    def update_role(self, role_qualified_id):
+        user_current_role = self.user.get_role(domain=self.user_domain)
+        self.role_updated = not (user_current_role
+                                 and user_current_role.get_qualified_id() == role_qualified_id)
+        if self.role_updated:
+            self.user.set_role(self.user_domain, role_qualified_id)
+
+    def save_log(self):
+        # Tracking for role is done post save to have role setup correctly on save
+        if self.role_updated:
+            new_role = self.user.get_role(domain=self.user_domain)
+            if new_role:
+                self.logger.add_info(_("Role: {new_role_name}[{new_role_id}]").format(
+                    new_role_name=new_role.name,
+                    new_role_id=new_role.get_id
+                ))
+            else:
+                self.logger.add_change_message("Role: None")
+
+        self._include_user_data_changes()
+        # ToDo: save log before saving user
+        self.logger.save()
+
+    def _include_user_data_changes(self):
+        # ToDo: consider putting just the diff
```

Comment:
either you should do this or remove the TODO (unless you plan to do it in a follow up PR?)

Reply:
I have not been able to find a good way to diff for this, so i would like to keep this note and take it in a follow up PR, in case I am not able to get to it before merging the base branch.
thanks for checking on this.